From 98f796edbf31e79340d6e6f2d0fcf4747c327a5d Mon Sep 17 00:00:00 2001
From: Erik Massop <e.massop@hccnet.nl>
Date: Thu, 19 Jun 2014 01:12:26 +0200
Subject: [PATCH 3/5] Improve detection of sse2 support

Squashed commit of the following:

commit af5b8a1264e1c82c8284298046d9d55f139e0169
Author: thome <thome@e5c1114b-a573-4582-9dac-f72f410959ce>
Date:   Mon Oct 14 11:53:48 2013 +0000

    improve sse-2 check

    git-svn-id: svn://scm.gforge.inria.fr/svnroot/gf2x/trunk@160 e5c1114b-a573-4582-9dac-f72f410959ce

commit b6e5b6e32f7605cdb69edeea60310f5adc6085b3
Author: thome <thome@e5c1114b-a573-4582-9dac-f72f410959ce>
Date:   Tue Sep 24 09:02:20 2013 +0000

    one more try.

    git-svn-id: svn://scm.gforge.inria.fr/svnroot/gf2x/trunk@159 e5c1114b-a573-4582-9dac-f72f410959ce

commit 9838b012dc79a5b30b49c142dd699fe5efb25165
Author: thome <thome@e5c1114b-a573-4582-9dac-f72f410959ce>
Date:   Tue Sep 24 07:35:49 2013 +0000

    gcc-4.8 + outsmarts our sse-2 test with -O4. Fix this.

    This is in fact rather problematic, since constant folding is very likely
    to affect many of our instruction set tests in this way.

    git-svn-id: svn://scm.gforge.inria.fr/svnroot/gf2x/trunk@158 e5c1114b-a573-4582-9dac-f72f410959ce

commit ed9c45dff31c216b97c64361353cbe7eb735e04f
Author: thome <thome@e5c1114b-a573-4582-9dac-f72f410959ce>
Date:   Mon Sep 23 21:57:23 2013 +0000

    Better sse-2 test

    git-svn-id: svn://scm.gforge.inria.fr/svnroot/gf2x/trunk@156 e5c1114b-a573-4582-9dac-f72f410959ce
---
 config/acinclude.m4 | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/config/acinclude.m4 b/config/acinclude.m4
index 9830d93..7aa51e7 100644
--- a/config/acinclude.m4
+++ b/config/acinclude.m4
@@ -74,10 +74,26 @@ AC_DEFUN([VERIFY_WORDSIZE],[
     esac
 ])
 
+# pepper this check with more sse-2 only statements, or we might be
+# fooled by some early athlon64 cpus supporting extended 3dnow, which
+# includes a subset of sse-2, but do not support the full sse-2 insn set.
 AC_DEFUN([SSE2_EXAMPLE],[AC_LANG_SOURCE([
 #include <emmintrin.h>
-__v2di x;
-int main() {}
+__v2di x;       /* Our code currently uses these, but it should not */
+int main(int argc, char * argv[]) {
+     __m128i foo = _mm_setr_epi32(argc, argc + 1, argc + 2, argc + 3);
+     __m128i bar = _mm_setr_epi32(argc + 3, argc + 2, argc + 1, argc);
+    foo = _mm_mullo_epi16(foo, bar);
+    foo = _mm_slli_epi64(foo, 1);
+    foo = _mm_xor_si128(bar, _mm_unpacklo_epi32 (foo, bar));
+    foo = _mm_srli_epi64(foo, 1);
+    foo = _mm_mullo_epi16(foo, bar);
+    foo = _mm_shuffle_epi32(foo, 78);
+    foo = _mm_xor_si128(bar, _mm_unpacklo_epi32 (foo, bar));
+    foo = _mm_srli_si128(foo, 1);
+
+    return _mm_extract_epi16(foo, 0) & (argc - 1);
+}
 ])])
 
 # Check whether we need some flag such as -msse2 in order to enable sse-2
-- 
2.0.0

